daal4py/sklearn/neighbors/knn.py

```diff
@@ -0,0 +1,525 @@
+
+            chunked_results = list(pairwise_distances_chunked(
+                X, self._fit_X, reduce_func=reduce_func,
+                metric=self.effective_metric_, n_jobs=n_jobs,
+                **kwds))
+
+        elif self._fit_method in ['ball_tree', 'kd_tree']:
+            if issparse(X):
+                raise ValueError(
+                    "%s does not work with sparse matrices. Densify the data, "
+                    "or set algorithm='brute'" % self._fit_method)
+            old_joblib = (
+                    LooseVersion(joblib.__version__) < LooseVersion('0.12'))
+            if old_joblib:
+                # Deal with change of API in joblib
+                check_pickle = False if old_joblib else None
+                delayed_query = delayed(_tree_query_parallel_helper,
+                                        check_pickle=check_pickle)
+                parallel_kwargs = {"backend": "threading"}
+            else:
+                delayed_query = delayed(_tree_query_parallel_helper)
+                parallel_kwargs = {"prefer": "threads"}
+            chunked_results = Parallel(n_jobs, **parallel_kwargs)(
+                delayed_query(
+                    self._tree, X[s], n_neighbors, return_distance)
+                for s in gen_even_slices(X.shape[0], n_jobs)
+            )
+        else:
+            raise ValueError("internal: _fit_method not recognized")
+
+        if chunked_results is not None:
+            if return_distance:
+                neigh_dist, neigh_ind = zip(*chunked_results)
+                results = np.vstack(neigh_dist), np.vstack(neigh_ind)
+            else:
+                results = np.vstack(chunked_results)
+
+        if not query_is_train:
+            return results
+        else:
+            # If the query data is the same as the indexed data, we would like
+            # to ignore the first nearest neighbor of every sample, i.e
+            # the sample itself.
+            if return_distance:
+                neigh_dist, neigh_ind = results
+            else:
+                neigh_ind = results
+
+            n_queries, _ = X.shape
+            sample_range = np.arange(n_queries)[:, None]
+            sample_mask = neigh_ind != sample_range
+
+            # Corner case: When the number of duplicates are more
+            # than the number of neighbors, the first NN will not
+            # be the sample, but a duplicate.
+            # In that case mask the first duplicate.
+            dup_gr_nbrs = np.all(sample_mask, axis=1)
+            sample_mask[:, 0][dup_gr_nbrs] = False
+            neigh_ind = np.reshape(
+                neigh_ind[sample_mask], (n_queries, n_neighbors - 1))
+
+            if return_distance:
+                neigh_dist = np.reshape(
+                    neigh_dist[sample_mask], (n_queries, n_neighbors - 1))
+                return neigh_dist, neigh_ind
+            return neigh_ind
+
+    def kneighbors_graph(self, X=None, n_neighbors=None,
```

Comment:
If we don't patch this function, why copy the content sklearn?

Reply:
Deleted